BSP/STM32F413/Src/BSP_CAN.c

```diff
@@ -0,0 +1,402 @@
+    CAN_FilterInitStruct.CAN_FilterMaskIdLow = 0x0000;
+    CAN_FilterInitStruct.CAN_FilterFIFOAssignment = 0;
+    CAN_FilterInitStruct.CAN_FilterActivation = ENABLE;
+    CAN_FilterInit(CAN3, &CAN_FilterInitStruct);
+
+    //CAN_SlaveStartBank(CAN1, 0);
+
+    /* Transmit Structure preparation */
+    gTxMessage[1].ExtId = 0x5;
+    gTxMessage[1].RTR = CAN_RTR_DATA;
+    gTxMessage[1].IDE = CAN_ID_STD;
+    gTxMessage[1].DLC = 1;
+
+    /* Receive Structure preparation */
+    gRxMessage[1].StdId = 0x00;
+    gRxMessage[1].ExtId = 0x00;
+    gRxMessage[1].IDE = CAN_ID_STD;
+    gRxMessage[1].DLC = 0;
+    gRxMessage[1].FMI = 0;
+
+    /* Enable FIFO 0 message pending Interrupt */
+    CAN_ITConfig(CAN3, CAN_IT_FMP0, ENABLE);
+
+    // Enable Rx interrupts
+    NVIC_InitStruct.NVIC_IRQChannel = CAN3_RX0_IRQn;
+    NVIC_InitStruct.NVIC_IRQChannelPreemptionPriority = 0x00;
+    NVIC_InitStruct.NVIC_IRQChannelSubPriority = 0x00;
+    NVIC_InitStruct.NVIC_IRQChannelCmd = ENABLE;
+    NVIC_Init(&NVIC_InitStruct);	
+
+}
+
+/**
+ * @brief   Transmits the data onto the CAN bus with the specified id
+ * @param   id : Message of ID. Also indicates the priority of message. The lower the value, the higher the priority.
+ * @param   data : data to be transmitted. The max is 8 bytes.
+ * @param   length : num of bytes of data to be transmitted. This must be <= 8 bytes or else the rest of the message is dropped.
+ * @return  ERROR if module was unable to transmit the data onto the CAN bus. SUCCESS indicates data was transmitted.
+ */
+ErrorStatus BSP_CAN_Write(CAN_t bus, uint32_t id, uint8_t data[8], uint8_t length) {
+    
+    gTxMessage[bus].StdId = id;
+    gTxMessage[bus].DLC = length;
+	for(int i = 0; i < length; i++){
+        gTxMessage[bus].Data[i] = data[i];
+    }
+
+    uint8_t retVal = (CAN_Transmit(bus == CAN_1 ? CAN1 : CAN3, &gTxMessage[bus]) != 0);
+    if (retVal == CAN_TxStatus_NoMailBox) {
+        return ERROR;
+    }
+    return SUCCESS;
+}
+
+/**
+ * @brief   Gets the data that was received from the CAN bus.
+ * @note    Non-blocking statement
+ * @pre     The data parameter must be at least 8 bytes or hardfault may occur.
+ * @param   id : pointer to store id of the message that was received.
+ * @param   data : pointer to store data that was received. Must be 8bytes or bigger.
+ * @return  ERROR if nothing was received so ignore id and data that was received. SUCCESS indicates data was received and stored.
+ */
+ErrorStatus BSP_CAN_Read(CAN_t bus, uint32_t *id, uint8_t *data) {
+    // If the queue is empty, return err
+    if(msg_queue_is_empty(&gRxQueue[bus])) {
+        return ERROR;
+    }
+    
+    // Get the message
+    msg_t msg;
+    msg_queue_get(&gRxQueue[bus], &msg);
+
+    // Transfer the message to the provided pointers
+    for(int i = 0; i < 8; i++){
+        data[i] = msg.data[i];
+    }
+    *id = msg.id;
+
+    return SUCCESS;
+}
+
+void CAN3_RX0_IRQHandler() {
+    #ifdef RTOS
```

Comment:
Are y'all defining `RTOS` anywhere? Not sure if y'all copied that part of the makefile from BPS or not